### ISSUE TITLE
Discrepancy between torch.nn.functional.linear and manual computation with bias under torch.autocast in torch.float16

### ISSUE THREAD

**Issue by TheoBoyer** [pytorch/pytorch#137080]
### 🐛 Describe the bug

When using torch.autocast with `dtype=torch.float16`, I observed a discrepancy between the outputs of `torch.nn.functional.linear(x, weight, bias)` and the manual computation `(x @ weight.T + bias)`.
```python
batch_size, dim = 32, 2048
device = 'cuda' if torch.cuda.is_available() else 'cpu'
x = torch.rand(batch_size, dim, device=device)
weight = torch.rand(dim, dim, device=device)
bias = torch.rand(dim, device=device)
with torch.autocast(device_type=device, dtype=torch.float16):
    linear = torch.nn.functional.linear(x, weight, bias)
    manual = (x @ weight.T + bias)
    print((linear - manual).abs().max().item()) # 0.505126953125
```

This difference increases linearly with respect to `dim`
![image](https://github.com/user-attachments/assets/56ad4300-acbc-417a-9472-aaca550c2102)
[Code to Reproduce the figure](https://gist.github.com/TheoBoyer/714a3355632afa92ae57c4ad2c2e6c9c)

Notably, when removing the bias, both methods produce identical results.
```python
batch_size, dim = 32, 2048
device = 'cuda' if torch.cuda.is_available() else 'cpu'
x = torch.rand(batch_size, dim, device=device)
weight = torch.rand(dim, dim, device=device)
with torch.autocast(device_type=device, dtype=torch.float16):
    linear = torch.nn.functional.linear(x, weight)
    manual = (x @ weight.T)
    print((linear - manual).abs().max().item()) # 0.0
```

### Versions

PyTorch version: 2.4.1+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.30.3
Libc version: glibc-2.35

Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.1.85+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 535.104.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               2
On-line CPU(s) list:                  0,1
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                           6
Model:                                79
Thread(s) per core:                   2
Core(s) per socket:                   1
Socket(s):                            1
Stepping:                             0
BogoMIPS:                             4399.99
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            32 KiB (1 instance)
L1i cache:                            32 KiB (1 instance)
L2 cache:                             256 KiB (1 instance)
L3 cache:                             55 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0,1
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Vulnerable; SMT Host state unknown
Vulnerability Meltdown:               Vulnerable
Vulnerability Mmio stale data:        Vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Vulnerable
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:             Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Vulnerable (Syscall hardening enabled)
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.12.1
[pip3] torch==2.4.1+cu121
[pip3] torchaudio==2.4.1+cu121
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.19.1+cu121
[conda] Could not collect

**Comment by albanD** [comment]
Hi,

This is most likely due to the fast that linear() is a single op where we can provide higher precision for the intermediary results. We cannot do that when you ask for each intermediary.

**Comment by TheoBoyer** [comment]
Hi,
Thank you for your response.
Initially i thought too but what confused me is that if you remove the bias, then the computation difference is `0.0`.
I had time to investigate further since, and found a better example to illustrate my point:
```python
import torch
batch_size, dim = 32, 2048
device = 'cuda' if torch.cuda.is_available() else 'cpu'
x = torch.randn(batch_size, dim, device=device)
weight = torch.randn(dim, dim, device=device)
zero_bias = torch.zeros(dim, device=device)
with torch.autocast(device_type=device, dtype=torch.float16):
    with_zero_bias = torch.nn.functional.linear(x, weight, zero_bias)
    without_bias = torch.nn.functional.linear(x, weight)
print((with_zero_bias - without_bias).abs().max().item())  # sometimes 0.125 sometimes 0.0625
```
In this case the difference between providing no bias and providing bias with zeros can produce a delta up to `0.125` on normally distributed inputs which seems big even in `float16`.

**Comment by albanD** [comment]
I guess you're running on cuda? Running on cpu locally gives me 0.0

**Comment by nikitaved** [comment]
I confirm I suffer from the same issue on `2.7.0.dev20250227+cu128` for `float32`, autocast or not autocast, the diff is about `1e-4` which is, imho, too high.
